python/prepare_release.py

```diff
@@ -0,0 +1,265 @@
+CITUS_CONTROL_SEARCH_PATTERN = r"^default_version*"
+
+MULT_EXT_MAJOR_SEARCH_PATTERN = r"^\s*\d{1,2}\.\d{1,2}devel$"
+MULT_EXT_PATCH_SEARCH_PATTERN = r"^\s*\d{1,2}\.\d{1,2}.\d{1,2}$"
+
+CONFIGURE_IN_SEARCH_PATTERN = "AC_INIT*"
+REPO_OWNER = "citusdata"
+
+
+class ResourceStatus(Enum):
+    INITIAL = 1
+    RELEASE_BRANCH_LOCAL = 2
+    RELEASE_BRANCH_REMOTE = 3
+    UPCOMING_VERSION_LOCAL = 4
+    UPCOMING_VERSION_REMOTE = 5
+    PULL_REQUEST_CREATED = 6
+
+
+def update_release(github_token: str, project_name: str, project_version: is_version(str), main_branch: str,
+                   earliest_pr_date: datetime, exec_path: str, is_test: bool = False,
+                   cherry_pick_enabled: bool = False) -> Tuple[str, str, str, ResourceStatus]:
+    multi_extension_sql_path = f"{exec_path}/{MULTI_EXTENSION_SQL}"
+    citus_control_file_path = f"{exec_path}/{CITUS_CONTROL}"
+    multi_extension_out_path = f"{exec_path}/{MULTI_EXTENSION_OUT}"
+    configure_in_path = f"{exec_path}/{CONFIGURE_IN}"
+    config_py_path = f"{exec_path}/{CONFIG_PY}"
+    distributed_dir_path = f"{exec_path}/{DISTRIBUTED_DIR_PATH}"
+
+    resource_status = ResourceStatus.INITIAL
+
+    project_version_details = get_version_details(project_version)
+    default_upcoming_version = f'{project_version_details["major"]}.{project_version_details["minor"]}.' \
+                               f'{str(int(project_version_details["patch"]) + 1)}'
+    upcoming_version = default_upcoming_version if os.getenv("UPCOMING_VERSION") is None else os.getenv(
+        "UPCOMING_VERSION")
+    upcoming_version_details = get_version_details(upcoming_version)
+    upcoming_patch_version = f'{upcoming_version_details["major"]}.{upcoming_version_details["minor"]}'
+    devel_version = f"{upcoming_patch_version}devel"
+    release_branch_name = f'release-{project_version_details["major"]}.{project_version_details["minor"]}'
+    release_branch_name = f"{release_branch_name}-test" if is_test else release_branch_name
+    g = Github(github_token)
+    repository = g.get_repo(f"{REPO_OWNER}/{project_name}")
+    newly_created_sql_file = ""
+    upcoming_version_branch = f"master-update-version-{uuid.uuid4()}"
+    # if major release
+    if is_major_release(project_version):
+        print(f"### {project_version} is a major release. Executing Major release flow### ")
+        # create release-X-Y branch
+        print(f"### Preparing {release_branch_name}...### ")
+        # checkout master
+        print(f"### Checking out {main_branch}...### ")
+        run(f"git checkout {main_branch}")
+        run(f"git pull")
+        print(f"### OK {main_branch} checked out and pulled### ")
+        # create release branch in release-X.Y format
+        print(f"### Creating release branch with name {release_branch_name}...### ")
+        run(f'git checkout -b {release_branch_name}')
+        print(f"### OK {release_branch_name} created### ")
+        resource_status = ResourceStatus.RELEASE_BRANCH_LOCAL
```

Comment:
>I think any code-blocks starting with a comment can go into separate helper functions. I won't put comments to all such things but I will give some examples for that in below comments.

e.g 1

Reply:
There is not much logic in this method If I group and make method for each of them, readability will be hurt. Most of them are 3 lines and two of them are releated to logging start and finish

Reply:
Fixed by https://github.com/citusdata/tools/pull/91/commits/a2e8aef98e41c785588ddb62c2ed0f902bee180d